forgerock-openbanking-uk-aspsp-as/forgerock-openbanking-uk-aspsp-as-gateway/forgerock-openbanking-uk-aspsp-as-gateway-server/src/main/java/com/forgerock/openbanking/aspsp/as/service/apiclient/ApiClientIdentityFactory.java

```diff
@@ -0,0 +1,224 @@
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+import javax.security.auth.x500.X500Principal;
+import java.security.Principal;
+import java.security.cert.CertificateParsingException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+@Slf4j
+public class ApiClientIdentityFactory {
+
+    private final String OBIE_ISSUER_NAME = "CN=OpenBanking Pre-Production Issuing CA,O=OpenBanking,C=GB";
+
+    public ApiClientIdentity getApiClientIdentity(Principal principal) throws ApiClientException, OAuth2InvalidClientException {
+
+        ApiClientIdentity apiClientIdentity = null;
+        if (principal instanceof PSD2Authentication) {
+            PSD2Authentication authentication = (PSD2Authentication) principal;
+            Psd2CertInfo certInfo = authentication.getPsd2CertInfo();
+            if(certInfo.isPsd2Cert()){
+                ApiClientCertificateType certType = getApiClientCertificateTypeFromPSD2(authentication);
+                switch (certType) {
+                    case FR_TRANSPORT:
+                        apiClientIdentity = new ApiClientIdentityFRTransport(authentication);
+                        break;
+                    case OBWAC:
+                        apiClientIdentity = new ApiClientIdentityOBWac(authentication);
+                        break;
+                    case QWAC:
+                        apiClientIdentity = new ApiClientIdentityQWac(authentication);
+                        break;
+                    default:
+                        String errorString = "Client presented an invalid Certificate " +
+                                "Type for use as a Transport certificate. Type presented ': " + certType + "'";
+                        log.debug("getApiClientIdentity() {}", errorString);
+                        throw new ApiClientException(errorString);
+
+                }
+            } else {
+                apiClientIdentity = createOBTransportIdentity(authentication);
+            }
+        } else if (principal instanceof X509Authentication){
+            X509Authentication authentication = (X509Authentication) principal;
+            apiClientIdentity = createOBTransportIdentity(authentication);
+        }
+        return apiClientIdentity;
+    }
+
+    private ApiClientIdentity createOBTransportIdentity(X509Authentication authentication)
+            throws OAuth2InvalidClientException, ApiClientException {
+        ApiClientCertificateType certType = getApiClientCertificateTypeFromX509(authentication);
+        if (certType == ApiClientCertificateType.OB_TRANSPORT) {
+            return new ApiClientIdentityOBTransport(authentication);
+        }
+        String errorString = "Client presented an invalid Certificate " +
+                "Type for use as a Transport certificate. Type presented ': " + certType + "'";
+        log.debug("getApiClientIdentity() {}", errorString);
```

Comment:
Could be better log.error or log.warn instead of log.debug? because contains the error message.